### PR TITLE
Add Smoke Test hook

### PR DESCRIPTION
**Pull request by reset**
This change adds a Smoke Test hook to the HookTable. It is currently
not executed because further refactorings are necessary to ensure
that hooks are run asynchronously without causing the main loop of
the manager to block.

This commit also contains a number of refactors to move hooks from
the package module in the supervisor crate to the service module.
While hook templates are defined in a package, they are largely a
component of the runtime and should be owned by the service module.

In a future commit we will hook up the smoke test and health check
hooks to background threads which will feed a cache on the service
struct to be read by the gossip server and http-gateway.

gif-keyboard-1669342386790492254

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/sup/src/manager/service/health.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum QuickCheck {
```
Comment (adamhjk) [inline components/sup/src/manager/service/health.rs]: Is this the right name? QuickCheck to me is the testing method.

**Comment by adamhjk** [comment]
gif-keyboard-124268809188495092

File: components/sup/src/manager/service/health.rs (hunk 1 above)
Comment (reset) [inline components/sup/src/manager/service/health.rs]: Yeah I wasn't sure - I was wavering between the obvious HealthCheck and this name. My reasoning now looking over things isn't good enough, I'm going to change this to HealthCheck

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/199513506).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/199513506).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**